File: crunch-scrunch/src/main/scala/org/apache/crunch/scrunch/PCollection.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.crunch.scrunch

import scala.collection.JavaConversions

import org.apache.crunch.{PCollection => JCollection, Pair => CPair, _}
import org.apache.crunch.lib.{Aggregate, Cartesian, Sample}
import org.apache.crunch.scrunch.Conversions._
import org.apache.crunch.scrunch.interpreter.InterpreterRunner
import org.apache.hadoop.mapreduce.TaskInputOutputContext
import org.apache.crunch.types.PType
import org.apache.crunch.fn.IdentityFn

class PCollection[S](val native: JCollection[S]) extends PCollectionLike[S, PCollection[S], JCollection[S]]
  with Incrementable[PCollection[S]] {

  import PCollection._

  type FunctionType[T] = S => T
  type CtxtFunctionType[T] = (S, TIOC) => T

  protected def wrapFlatMapFn[T](fmt: S => TraversableOnce[T]) = flatMapFn(fmt)
  protected def wrapMapFn[T](fmt: S => T) = mapFn(fmt)
  protected def wrapFilterFn(fmt: S => Boolean) = filterFn(fmt)
  protected def wrapFlatMapWithCtxtFn[T](fmt: (S, TIOC) => TraversableOnce[T]) = flatMapWithCtxtFn(fmt)
  protected def wrapMapWithCtxtFn[T](fmt: (S, TIOC) => T) = mapWithCtxtFn(fmt)
  protected def wrapFilterWithCtxtFn(fmt: (S, TIOC) => Boolean) = filterWithCtxtFn(fmt)
  protected def wrapPairFlatMapFn[K, V](fmt: S => TraversableOnce[(K, V)]) = pairFlatMapFn(fmt)
  protected def wrapPairMapFn[K, V](fmt: S => (K, V)) = pairMapFn(fmt)

  def withPType(pt: PType[S]): PCollection[S] = {
    val ident: MapFn[S, S] = IdentityFn.getInstance()
    wrap(native.parallelDo("withPType", ident, pt))
  }

  def union(others: PCollection[S]*) = {
    new PCollection[S](native.union(others.map(_.native) : _*))
  }

  def by[K: PTypeH](f: S => K): PTable[K, S] = {
    val ptype = getTypeFamily().tableOf(implicitly[PTypeH[K]].get(getTypeFamily()), native.getPType())
    parallelDo(mapKeyFn[S, K](f), ptype)
  }

  def groupBy[K: PTypeH](f: S => K): PGroupedTable[K, S] = {
    by(f).groupByKey
  }

  def cross[S2](other: PCollection[S2]): PCollection[(S, S2)] = {
    val inter = Cartesian.cross(this.native, other.native)
    val f = (in: CPair[S, S2]) => (in.first(), in.second())
    inter.parallelDo(mapFn(f), getTypeFamily().tuple2(pType, other.pType))
  }

  def materialize() = {
    setupRun()
    JavaConversions.iterableAsScalaIterable[S](native.materialize)
  }

  protected def wrap(newNative: JCollection[_]) = new PCollection[S](newNative.asInstanceOf[JCollection[S]])

  def collect[T, To](pf: PartialFunction[S, T])(implicit pt: PTypeH[T], b: CanParallelTransform[T, To]) = {
    filter(pf.isDefinedAt(_)).map(pf)(pt, b)
  }

  def increment(groupName: String, counterName: String, count: Long) = {
    new IncrementPCollection[S](this).apply(groupName, counterName, count)
  }

  def incrementIf(f: S => Boolean) = new IncrementIfPCollection[S](this, f)

  def count() = {
    val count = new PTable[S, java.lang.Long](Aggregate.count(native))
    count.mapValues(_.longValue())
  }

  def max()(implicit converter: Converter[S, S]) = {
    setupRun()
    PObject(Aggregate.max(native))(converter)
  }

  def min()(implicit converter: Converter[S, S]) = {
    setupRun()
    PObject(Aggregate.min(native))(converter)
  }

  def sample(acceptanceProbability: Double) = {
    wrap(Sample.sample(native, acceptanceProbability))
  }

  def sample(acceptanceProbability: Double, seed: Long) = {
    wrap(Sample.sample(native, seed, acceptanceProbability))
  }

  def pType() = native.getPType()
}

trait SDoFn[S, T] extends DoFn[S, T] with Function1[S, TraversableOnce[T]] {
  override def process(input: S, emitter: Emitter[T]) {
    for (v <- apply(input)) {
      emitter.emit(v)
    }
  }
}

trait SMapFn[S, T] extends MapFn[S, T] with Function1[S, T] {
  override def map(input: S) = apply(input)
}

trait SFilterFn[T] extends FilterFn[T] with Function1[T, Boolean] {
  override def accept(input: T) = apply(input)
}

class SDoWithCtxtFn[S, T](val f: (S, TaskInputOutputContext[_, _, _, _]) => TraversableOnce[T]) extends DoFn[S, T] {
  override def process(input: S, emitter: Emitter[T]) {
    for (v <- f(input, getContext)) {
      emitter.emit(v)
    }
  }
}

class SMapWithCtxtFn[S, T](val f: (S, TaskInputOutputContext[_, _, _, _]) => T) extends MapFn[S, T] {
  override def map(input: S) = f(input, getContext)
}

class SFilterWithCtxtFn[T](val f: (T, TaskInputOutputContext[_, _, _, _]) => Boolean) extends FilterFn[T] {
  override def accept(input: T) = f.apply(input, getContext)
}

trait SDoPairFn[S, K, V] extends DoFn[S, CPair[K, V]] with Function1[S, TraversableOnce[(K, V)]] {
  override def process(input: S, emitter: Emitter[CPair[K, V]]) {
    for (v <- apply(input)) {
      emitter.emit(CPair.of(v._1, v._2))
    }
  }
}

trait SMapPairFn[S, K, V] extends MapFn[S, CPair[K, V]] with Function1[S, (K, V)] {
  override def map(input: S): CPair[K, V] = {
    val t = apply(input)
    CPair.of(t._1, t._2)
  }
}

trait SMapKeyFn[S, K] extends MapFn[S, CPair[K, S]] with Function1[S, K] {
  override def map(input: S): CPair[K, S] = {
    CPair.of(apply(input), input)
  }
}

object PCollection {
  type TIOC = TaskInputOutputContext[_, _, _, _]

  def flatMapFn[S, T](fn: S => TraversableOnce[T]) = {
    new SDoFn[S, T] { def apply(s: S) = fn(s) }
  }

  def mapFn[S, T](fn: S => T) = {
    new SMapFn[S, T] { def apply(s: S) = fn(s) }
  }

  def filterFn[S](fn: S => Boolean) = {
    new SFilterFn[S] { def apply(x: S) = fn(x) }
  }

  def flatMapWithCtxtFn[S, T](fn: (S, TIOC) => TraversableOnce[T]) = {
    new SDoWithCtxtFn[S, T](fn)
  }

  def mapWithCtxtFn[S, T](fn: (S, TIOC) => T) = {
    new SMapWithCtxtFn[S, T](fn)
  }

  def filterWithCtxtFn[S](fn: (S, TIOC) => Boolean) = {
    new SFilterWithCtxtFn[S](fn)
  }

  def mapKeyFn[S, K](fn: S => K) = {
    new SMapKeyFn[S, K] { def apply(x: S) = fn(x) }
  }

  def pairMapFn[S, K, V](fn: S => (K, V)) = {
    new SMapPairFn[S, K, V] { def apply(s: S) = fn(s) }
  }

  def pairFlatMapFn[S, K, V](fn: S => TraversableOnce[(K, V)]) = {
    new SDoPairFn[S, K, V] { def apply(s: S) = fn(s) }
  }
}

File: crunch-core/src/main/java/org/apache/crunch/PCollection.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.crunch;

import java.util.Collection;

import org.apache.crunch.types.PTableType;
import org.apache.crunch.types.PType;
import org.apache.crunch.types.PTypeFamily;

/**
 * A representation of an immutable, distributed collection of elements that is
 * the fundamental target of computations in Crunch.
 *
 */
public interface PCollection<S> {
  /**
   * Returns the {@code Pipeline} associated with this PCollection.
   */
  Pipeline getPipeline();

  /**
   * Returns a {@code PCollection} instance that acts as the union of this
   * {@code PCollection} and the given {@code PCollection}.
   */
  PCollection<S> union(PCollection<S> other);
  
  /**
   * Returns a {@code PCollection} instance that acts as the union of this
   * {@code PCollection} and the input {@code PCollection}s.
   */
  PCollection<S> union(PCollection<S>... collections);

  /**
   * Applies the given doFn to the elements of this {@code PCollection} and
   * returns a new {@code PCollection} that is the output of this processing.
   *
   * @param doFn
   *          The {@code DoFn} to apply
   * @param type
   *          The {@link PType} of the resulting {@code PCollection}
   * @return a new {@code PCollection}
   */
  <T> PCollection<T> parallelDo(DoFn<S, T> doFn, PType<T> type);

  /**
   * Applies the given doFn to the elements of this {@code PCollection} and
   * returns a new {@code PCollection} that is the output of this processing.
   *
   * @param name
   *          An identifier for this processing step, useful for debugging
   * @param doFn
   *          The {@code DoFn} to apply
   * @param type
   *          The {@link PType} of the resulting {@code PCollection}
   * @return a new {@code PCollection}
   */
  <T> PCollection<T> parallelDo(String name, DoFn<S, T> doFn, PType<T> type);
  
  /**
   * Applies the given doFn to the elements of this {@code PCollection} and
   * returns a new {@code PCollection} that is the output of this processing.
   *
   * @param name
   *          An identifier for this processing step, useful for debugging
   * @param doFn
   *          The {@code DoFn} to apply
   * @param type
   *          The {@link PType} of the resulting {@code PCollection}
   * @param options
   *          Optional information that is needed for certain pipeline operations
   * @return a new {@code PCollection}
   */
  <T> PCollection<T> parallelDo(String name, DoFn<S, T> doFn, PType<T> type,
      ParallelDoOptions options);

  /**
   * Similar to the other {@code parallelDo} instance, but returns a
   * {@code PTable} instance instead of a {@code PCollection}.
   *
   * @param doFn
   *          The {@code DoFn} to apply
   * @param type
   *          The {@link PTableType} of the resulting {@code PTable}
   * @return a new {@code PTable}
   */
  <K, V> PTable<K, V> parallelDo(DoFn<S, Pair<K, V>> doFn, PTableType<K, V> type);

  /**
   * Similar to the other {@code parallelDo} instance, but returns a
   * {@code PTable} instance instead of a {@code PCollection}.
   *
   * @param name
   *          An identifier for this processing step
   * @param doFn
   *          The {@code DoFn} to apply
   * @param type
   *          The {@link PTableType} of the resulting {@code PTable}
   * @return a new {@code PTable}
   */
  <K, V> PTable<K, V> parallelDo(String name, DoFn<S, Pair<K, V>> doFn, PTableType<K, V> type);
  
  /**
   * Similar to the other {@code parallelDo} instance, but returns a
   * {@code PTable} instance instead of a {@code PCollection}.
   *
   * @param name
   *          An identifier for this processing step
   * @param doFn
   *          The {@code DoFn} to apply
   * @param type
   *          The {@link PTableType} of the resulting {@code PTable}
   * @param options
   *          Optional information that is needed for certain pipeline operations
   * @return a new {@code PTable}
   */
  <K, V> PTable<K, V> parallelDo(String name, DoFn<S, Pair<K, V>> doFn, PTableType<K, V> type,
      ParallelDoOptions options);

  /**
   * Similar to other instances of {@code parallelDo}, but designed for Java lambdas.
   */
  <T> PCollection<T> parallelDo(IDoFn<S, T> fn, PType<T> type);

  /**
   * Similar to other instances of {@code parallelDo}, but designed for Java lambdas.
   */
  <K, V> PTable<K, V> parallelDo(IDoFn<S, Pair<K, V>> fn, PTableType<K, V> type);

  /**
   * Similar to other instances of {@code parallelDo}, but designed for Java lambdas.
   */
  <T> PCollection<T> parallelDo(String name, IDoFn<S, T> fn, PType<T> type);

  /**
   * Similar to other instances of {@code parallelDo}, but designed for Java lambdas.
   */
  <K, V> PTable<K, V> parallelDo(String name, IDoFn<S, Pair<K, V>> fn, PTableType<K, V> type);

  /**
   * Similar to other instances of {@code parallelDo}, but designed for Java lambdas.
   */
  <T> PCollection<T> parallelDo(String name, IDoFn<S, T> fn, PType<T> type, ParallelDoOptions options);

  /**
   * Similar to other instances of {@code parallelDo}, but designed for Java lambdas.
   */
  <K, V> PTable<K, V> parallelDo(String name, IDoFn<S, Pair<K, V>> fn, PTableType<K, V> type, ParallelDoOptions options);

  /**
   * For each element of this {@code PCollection}, generate 0 to N output values using the
   * given {@code IFlatMapFn}. Designed for Java lambdas.
   */
  <T> PCollection<T> flatMap(IFlatMapFn<S, T> fn, PType<T> type);

  /**
   * For each element of this {@code PCollection}, generate 0 to N output values using the
   * given {@code IFlatMapFn}. Designed for Java lambdas.
   */
  <K, V> PTable<K, V> flatMap(IFlatMapFn<S, Pair<K, V>> fn, PTableType<K, V> type);

  /**
   * For each element of this {@code PCollection}, generate 0 to N output values using the
   * given {@code IFlatMapFn}. Designed for Java lambdas.
   */
  <T> PCollection<T> flatMap(String name, IFlatMapFn<S, T> fn, PType<T> type);

  /**
   * For each element of this {@code PCollection}, generate 0 to N output values using the
   * given {@code IFlatMapFn}. Designed for Java lambdas.
   */
  <K, V> PTable<K, V> flatMap(String name, IFlatMapFn<S, Pair<K, V>> fn, PTableType<K, V> type);

  /**
   * For each element of this {@code PCollection}, generate one output value using the
   * given {@code IMapFn}. Designed for Java lambdas.
   */
  <T> PCollection<T> map(IMapFn<S, T> fn, PType<T> type);

  /**
   * For each element of this {@code PCollection}, generate one output value using the
   * given {@code IMapFn}. Designed for Java lambdas.
   */
  <K, V> PTable<K, V> map(IMapFn<S, Pair<K, V>> fn, PTableType<K, V> type);

  /**
   * For each element of this {@code PCollection}, generate one output value using the
   * given {@code IMapFn}. Designed for Java lambdas.
   */
  <T> PCollection<T> map(String name, IMapFn<S, T> fn, PType<T> type);

  /**
   * For each element of this {@code PCollection}, generate one output value using the
   * given {@code IMapFn}. Designed for Java lambdas.
   */
  <K, V> PTable<K, V> map(String name, IMapFn<S, Pair<K, V>> fn, PTableType<K, V> type);

  /**
   * Filter elements of this {@code PCollection} using the given {@code IFilterFn}.
   * Designed for Java lambdas.
   */
  PCollection<S> filter(IFilterFn<S> fn);

  /**
   * Filter elements of this {@code PCollection} using the given {@code IFilterFn}.
   * Designed for Java lambdas.
   */
  PCollection<S> filter(String name, IFilterFn<S> fn);

  /**
   * Write the contents of this {@code PCollection} to the given {@code Target},
   * using the storage format specified by the target.
   *
   * @param target
   *          The target to write to
   */
  PCollection<S> write(Target target);

  /**
   * Write the contents of this {@code PCollection} to the given {@code Target},
   * using the given {@code Target.WriteMode} to handle existing
   * targets.
   * 
   * @param target
   *          The target
   * @param writeMode
   *          The rule for handling existing outputs at the target location
   */
  PCollection<S> write(Target target, Target.WriteMode writeMode);
  
  /**
   * Returns a reference to the data set represented by this PCollection that
   * may be used by the client to read the data locally.
   */
  Iterable<S> materialize();

  /**
   * Marks this data as cached using the default {@link CachingOptions}. Cached {@code PCollection}s will only
   * be processed once, and then their contents will be saved so that downstream code can process them many times.
   *
   * @return this {@code PCollection} instance
   */
  PCollection<S> cache();

  /**
   * Marks this data as cached using the given {@code CachingOptions}. Cached {@code PCollection}s will only
   * be processed once and then their contents will be saved so that downstream code can process them many times.
   *
   * @param options the options that control the cache settings for the data
   * @return this {@code PCollection} instance
   */
  PCollection<S> cache(CachingOptions options);

  /**
   * @return A {@code PObject} encapsulating an in-memory {@link Collection} containing the values
   * of this {@code PCollection}.
   */
  PObject<Collection<S>> asCollection();

  /**
   * @return The first element of this {@code PCollection}.
   */
  PObject<S> first();

  /**
   * Adds the materialized data in this {@code PCollection} as a dependency to the given
   * {@code PipelineCallable} and registers it with the {@code Pipeline} associated with this
   * instance.
   *
   * @param label the label to use inside of the PipelineCallable for referencing this PCollection
   * @param pipelineCallable the function itself
   *
   * @return The value of the {@code getOutput} function on the given argument.
   */
  <Output> Output sequentialDo(String label, PipelineCallable<Output> pipelineCallable);

  /**
   * @return A reference to the data in this instance that can be read from a job running
   * on a cluster.
   *
   * @param materialize If true, materialize this data before returning a reference to it
   */
  ReadableData<S> asReadable(boolean materialize);

  /**
   * Returns the {@code PType} of this {@code PCollection}.
   */
  PType<S> getPType();

  /**
   * Returns the {@code PTypeFamily} of this {@code PCollection}.
   */
  PTypeFamily getTypeFamily();

  /**
   * Returns the size of the data represented by this {@code PCollection} in
   * bytes.
   */
  long getSize();

  /**
   * Returns the number of elements represented by this {@code PCollection}.
   *
   * @return An {@code PObject} containing the number of elements in this {@code PCollection}.
   */
  PObject<Long> length();

  /**
   * Returns a shorthand name for this PCollection.
   */
  String getName();

  /**
   * Apply the given filter function to this instance and return the resulting
   * {@code PCollection}.
   */
  PCollection<S> filter(FilterFn<S> filterFn);

  /**
   * Apply the given filter function to this instance and return the resulting
   * {@code PCollection}.
   *
   * @param name
   *          An identifier for this processing step
   * @param filterFn
   *          The {@code FilterFn} to apply
   */
  PCollection<S> filter(String name, FilterFn<S> filterFn);

  /**
   * Apply the given map function to each element of this instance in order to
   * create a {@code PTable}.
   */
  <K> PTable<K, S> by(MapFn<S, K> extractKeyFn, PType<K> keyType);

  /**
   * Apply the given {@code IMapFn} to each element of this instance in order to
   * create a {@code PTable}. Designed for use with Java 8 lambdas.
   */
  <K> PTable<K, S> by(IMapFn<S, K> extractKeyFn, PType<K> keyType);

  /**
   * Apply the given map function to each element of this instance in order to
   * create a {@code PTable}.
   *
   * @param name
   *          An identifier for this processing step
   * @param extractKeyFn
   *          The {@code MapFn} to apply
   */
  <K> PTable<K, S> by(String name, MapFn<S, K> extractKeyFn, PType<K> keyType);

  /**
   * Returns a {@code PTable} instance that contains the counts of each unique
   * element of this PCollection.
   */
  PTable<S, Long> count();

  /**
   * Returns a {@code PObject} of the maximum element of this instance.
   */
  PObject<S> max();

  /**
   * Returns a {@code PObject} of the minimum element of this instance.
   */
  PObject<S> min();
  
  /**
   * Returns a {@code PCollection} that contains the result of aggregating all values in this instance.
   */
  PCollection<S> aggregate(Aggregator<S> aggregator);
}
